### PR TITLE
[ScrollingTree mac] Fix scrolling when event region and scroll container are distinct

### PR DESCRIPTION
**Pull request by mdubet**
#### 2b7aa256562d1d2fa0bc22b53d9065a0d0556064
<pre>
[ScrollingTree mac] Fix scrolling when event region and scroll container are distinct
<a href="https://bugs.webkit.org/show_bug.cgi?id=258255">https://bugs.webkit.org/show_bug.cgi?id=258255</a>
<a href="https://rdar.apple.com/110954175">rdar://110954175</a>

Reviewed by Simon Fraser.

The current code (mac) assumes that the platform layer which reacts to
a scroll event is also the one containing the scroll container but it&apos;s not always the case
(like WPT example: when the scroll container has pointer-event:none but some child still reacts to event).

* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_hit_test_scroll_visible_descendant-expected.txt:
* Source/WebCore/page/scrolling/mac/ScrollingTreeMac.mm:
(ScrollingTreeMac::scrollingNodeForPoint):
* Source/WebCore/platform/graphics/cocoa/WebCoreCALayerExtras.h:
(WebCore::collectDescendantLayersAtPoint):
* Source/WebCore/platform/graphics/cocoa/WebCoreCALayerExtras.mm:
(WebCore::collectDescendantLayersAtPoint):
(WebCore::layersAtPointToCheckForScrolling):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
(WebKit::RemoteScrollingTreeMac::scrollingNodeForPoint):

Canonical link: <a href="https://commits.webkit.org/270094@main">https://commits.webkit.org/270094@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/c45b02f874d968beec2235d62e04a7ad8775494b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/24537 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/3088 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/25805 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/26665 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/22556 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/24807 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/4735 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/523 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/22950 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/24782 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/2149 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21199 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27252 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/1880 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/22125 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/28321 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/22376 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/22454 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26114 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/1808 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/138 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/3118 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5879 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/2266 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/2182 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | |

### CODE REVIEW

File: Source/WebCore/platform/graphics/cocoa/WebCoreCALayerExtras.h [hunk 1]
```diff
@@ -52,7 +52,7 @@
 namespace WebCore {
 
 using LayerAndPoint = std::pair<CALayer *, FloatPoint>;
-WEBCORE_EXPORT void collectDescendantLayersAtPoint(Vector<LayerAndPoint, 16>& layersAtPoint, CALayer *parent, CGPoint, const std::function<bool(CALayer *, CGPoint localPoint)>& pointInLayerFunction);
+WEBCORE_EXPORT void collectDescendantLayersAtPoint(Vector<LayerAndPoint, 16>& layersAtPoint, CALayer *parent, CGPoint, const std::optional<std::function<bool(CALayer *, CGPoint localPoint)>>& pointInLayerFunction = { });
```
Comment (cdumez) [inline Source/WebCore/platform/graphics/cocoa/WebCoreCALayerExtras.h]: `std::optional<std::function<>>` is rarely a good idea since std::function has a null state. You could just use `= nullptr` for the default value and then null-check the std::function before calling it.

File: Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm [hunk 2]
```diff
@@ -465,27 +465,56 @@ static bool layerEventRegionContainsPoint(CALayer *layer, CGPoint localPoint)
...
+            reversedLayersAtPointWithEventRegion.append({layer,point});
+    }
+
+    if (reversedLayersAtPointWithEventRegion.size()) {
+        auto* frontmostLayer = reversedLayersAtPointWithEventRegion.first().first;
+        if (!frontmostLayer)
+            return rootScrollingNode;
+
+        for (auto [layer, point]: reversedLayersAtPointWithEventRegion) {
+            auto scrollingNodeForLayer = [&] (auto layer, auto point) -> ScrollingTreeNode* {
```
Comment (cdumez) [inline Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm]: `RefPtr scrollingNodeForLayer`. Please avoid raw pointers in new code.

File: Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm [hunk 3]
```diff
@@ -465,27 +465,56 @@ static bool layerEventRegionContainsPoint(CALayer *layer, CGPoint localPoint)
...
+                auto nodeID = scrollingNodeIDForLayer(layer);
+
+                auto* scrollingNode = nodeForID(nodeID);
```
Comment (cdumez) [inline Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm]: RefPtr

File: Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm [hunk 4]
```diff
@@ -465,27 +465,56 @@ static bool layerEventRegionContainsPoint(CALayer *layer, CGPoint localPoint)
...
+                    LOG_WITH_STREAM(UIHitTesting, stream << "RemoteScrollingTreeMac " << this << " scrollingNodeForPoint " << point << " found scrolling node " << nodeID);
+                    return scrollingNode;
+                }
 
-            if (isScrolledBy(*this, nodeID, frontmostLayer)) {
-                LOG_WITH_STREAM(UIHitTesting, stream << "RemoteScrollingTreeMac " << this << " scrollingNodeForPoint " << point << " found scrolling node " << nodeID);
+                return nullptr;
+            };
+
+            if (auto scrollingNode = scrollingNodeForLayer(layer, point)) {
```
Comment (cdumez) [inline Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm]: No curly brackets for one lines, also `RefPtr`

File: Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm [hunk 5]
```diff
@@ -465,27 +465,56 @@ static bool layerEventRegionContainsPoint(CALayer *layer, CGPoint localPoint)
...
             }
+
+            // If we haven't find a scrolling node for this layer, try to find a scrolling node from a layer below this
+            auto reversedLayers = WTF::makeReversedRange(layersAtPoint);
+            auto startingLayer = reversedLayers.begin();
+            while (startingLayer->first != layer && startingLayer < reversedLayers.end())
+                startingLayer++;
+            auto parentLayer = startingLayer + 1;
+            while (parentLayer < reversedLayers.end()) {
+                if (auto scrollingNode = scrollingNodeForLayer(parentLayer->first, parentLayer->second))
```
Comment (cdumez) [inline Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm]: RefPtr

File: Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm (hunk 2 above)
Comment (cdumez) [reply Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm]: The lambda should likely return a `RefPtr< ScrollingTreeNode>` too since this is what it will have internally with the changes I suggest below.

File: Source/WebCore/platform/graphics/cocoa/WebCoreCALayerExtras.h (hunk 1 above)
Comment (mdubet) [reply Source/WebCore/platform/graphics/cocoa/WebCoreCALayerExtras.h]: TIL that `std::function` has a null state. 👍

**Comment by nt1m** [comment]
@mdubet Seems like mac-wk2 is still failing the newly added test.

File: Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm [hunk 6]
```diff
@@ -465,26 +465,54 @@ static bool layerEventRegionContainsPoint(CALayer *layer, CGPoint localPoint)
...
+            return rootScrollingNode;
+
+        for (auto [layer, point] : reversedLayersAtPointWithEventRegion) {
+            auto scrollingNodeForLayer = [&] (auto layer, auto point) -> RefPtr<ScrollingTreeNode> {
+                auto nodeID = scrollingNodeIDForLayer(layer);
 
-            auto* scrollingNode = nodeForID(nodeID);
-            if (!is<ScrollingTreeScrollingNode>(scrollingNode))
-                continue;
+                RefPtr<ScrollingTreeNode> scrollingNode = nodeForID(nodeID);
```
Comment (nt1m) [inline Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm]:
```suggestion
                RefPtr scrollingNode = nodeForID(nodeID);
```

File: Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm [hunk 7]
```diff
@@ -465,26 +465,54 @@ static bool layerEventRegionContainsPoint(CALayer *layer, CGPoint localPoint)
...
-                LOG_WITH_STREAM(UIHitTesting, stream << "RemoteScrollingTreeMac " << this << " scrollingNodeForPoint " << point << " found scrolling node " << nodeID);
+                if (isScrolledBy(*this, nodeID, frontmostLayer)) {
+                    LOG_WITH_STREAM(UIHitTesting, stream << "RemoteScrollingTreeMac " << this << " scrollingNodeForPoint " << point << " found scrolling node " << nodeID);
+                    return scrollingNode;
+                }
+
+                return nullptr;
+            };
+
+            if (RefPtr<ScrollingTreeNode> scrollingNode = scrollingNodeForLayer(layer, point))
```
Comment (nt1m) [inline Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm]:
```suggestion
            if (RefPtr scrollingNode = scrollingNodeForLayer(layer, point))
```

File: LayoutTests/platform/mac-ventura-wk2/TestExpectations [hunk 8]
```diff
@@ -32,5 +32,8 @@ webgl/1.0.x/conformance/glsl/misc/shader-with-non-reserved-words-3-of-8.html [ P
 
 webkit.org/b/261444 [ Debug x86_64 ] http/tests/security/referrer-policy-header.html [ Pass Timeout ]
 
+# AsyncOverflowScrolling in enabled on Sonoma+
```
Comment (smfr) [inline LayoutTests/platform/mac-ventura-wk2/TestExpectations]: That comment isn't accurate. What you mean to say is "UI-side compositing is only enabled on Sonoma+".

File: LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_hit_test_scroll_visible_descendant.html [hunk 9]
```diff
@@ -1,4 +1,4 @@
-<!doctype html>
+<!DOCTYPE html> <!-- webkit-test-runner [ AsyncOverflowScrollingEnabled=true ] -->
```
Comment (nt1m) [inline LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_hit_test_scroll_visible_descendant.html]: Please undo this change.

Comment (mdubet) [reply LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_hit_test_scroll_visible_descendant.html]: No, we need this to run EWS. It will be gone when we rebased after https://github.com/WebKit/WebKit/pull/19664 is merged

File: Source/WebCore/platform/graphics/cocoa/WebCoreCALayerExtras.h [hunk 10]
```diff
@@ -52,7 +52,24 @@
 namespace WebCore {
 
 using LayerAndPoint = std::pair<CALayer *, FloatPoint>;
-WEBCORE_EXPORT void collectDescendantLayersAtPoint(Vector<LayerAndPoint, 16>& layersAtPoint, CALayer *parent, CGPoint, const std::function<bool(CALayer *, CGPoint localPoint)>& pointInLayerFunction);
+WEBCORE_EXPORT void collectDescendantLayersAtPoint(Vector<LayerAndPoint, 16>& layersAtPoint, CALayer *parent, CGPoint, const std::function<bool(CALayer *, CGPoint localPoint)>& pointInLayerFunction = { });
+
+auto layersAtPointToCheckForScrolling(auto layerEventRegionContainsPoint, auto scrollingNodeIDForLayer, const auto& layer, const auto& point, bool& hasAnyNonInteractiveScrollingLayers)
```
Comment (smfr) [inline Source/WebCore/platform/graphics/cocoa/WebCoreCALayerExtras.h]: There's no reason to inline this function.

File: Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm [hunk 11]
```diff
@@ -467,28 +467,47 @@ static bool layerEventRegionContainsPoint(CALayer *layer, CGPoint localPoint)
...
 
-            auto* scrollingNode = nodeForID(nodeID);
-            if (!is<ScrollingTreeScrollingNode>(scrollingNode))
+            if (!layerEventRegionContainsPoint(layer, point))
                 continue;
 
-            if (isScrolledBy(*this, nodeID, frontmostLayer)) {
-                LOG_WITH_STREAM(UIHitTesting, stream << "RemoteScrollingTreeMac " << this << " scrollingNodeForPoint " << point << " found scrolling node " << nodeID);
+            auto scrollingNodeForLayer = [&] (auto layer, auto point) -> RefPtr<ScrollingTreeNode> {
+                auto nodeID = scrollingNodeIDForLayer(layer);
```
Comment (smfr) [inline Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:499]: Might as well throw in your `UNUSED_PARAM(point);` here too (since it's only used by logging).

**Review by smfr — APPROVED** [review]

File: LayoutTests/TestExpectations [hunk 12]
```diff
@@ -6843,6 +6843,9 @@ imported/w3c/web-platform-tests/scroll-animations/scroll-timelines/two-animation
 imported/w3c/web-platform-tests/scroll-animations/scroll-timelines/two-animations-attach-to-same-scroll-timeline.html [ Skip ]
 imported/w3c/web-platform-tests/scroll-animations/view-timelines/range-boundary.html [ Skip ]
 
+# EWS can't run this test because AsyncOverflowScrolling is disabled
+imported/w3c/web-platform-tests/pointerevents/pointerevent_hit_test_scroll_visible_descendant.html [ Skip ]
```
Comment (nt1m) [inline LayoutTests/TestExpectations]: That probably means no Interop points since wpt.fyi runs Ventura :(

Comment (mdubet) [reply LayoutTests/TestExpectations]: AsyncOverflowScrolling is not Sonoma only (UI-side compositing is, but we do both code paths in this PR so we should be ok for Interop).

Comment (nt1m) [reply LayoutTests/TestExpectations]: So why do we skip this test on Ventura? If this fails on Ventura, it'll fail on wpt.fyi

Comment (mdubet) [reply LayoutTests/TestExpectations]: We skip everywhere here. (because for this to run on EWS, we first need to activate AsyncOverflow which will be done in another PR)

File: Source/WebCore/platform/graphics/cocoa/WebCoreCALayerExtras.h (hunk 10 above)
Comment (mdubet) [reply Source/WebCore/platform/graphics/cocoa/WebCoreCALayerExtras.h]: It was templated to avoid writing the long type of the function args...
It's implemented in the .cpp now
